src/tutorial.rs

```diff
@@ -0,0 +1,358 @@
+//! networking fun. This library makes heavy use of asynchronous Rust. In case
+//! you are not familiar with the concepts the Rust
+//! [async-book](https://rust-lang.github.io/async-book/) should proof useful.
+//! People new to libp2p might prefer to get a general overview at libp2p.io
+//! first, though libp2p knowledge is not required for this tutorial.
+//!
+//! We are going to build a small `ping` clone, sending a ping to a peer,
+//! expecting a pong as a response.
+//!
+//! ## Scaffolding
+//!
+//! Let's start off by
+//!
+//! 1. Creating a new crate: `cargo init rust-libp2p-tutorial`
+//!
+//! 2. And adding `libp2p` as well as `futures` as a dependency in the
+//!    `Cargo.toml` file:
+//!
+//!    ```yaml
+//!    [package]
+//!        name = "rust-libp2p-tutorial"
+//!        version = "0.1.0"
+//!        authors = ["Max Inden <mail@max-inden.de>"]
+//!        edition = "2018"
+//!
+//!    [dependencies]
+//!        libp2p = "<insert-current-version-here>"
+//!        futures = "<insert-current-version-here>"
+//!    ```
+//!
+//! ## Network identity
+//!
+//! With all the scaffolding in place, we can dive into the libp2p specifics. At
+//! first we need to create a network identity for our local node in `fn
+//! main()`. Identities in libp2p are handled via a public and private key pair.
+//! Nodes identify each other via their [`PeerId`](crate::PeerId) which is
+//! derived from the public key.
+//!
+//! ```rust
+//! use libp2p::{identity, PeerId};
+//! use std::error::Error;
+//!
+//! fn main() -> Result<(), Box<dyn Error>> {
+//!     let local_key = identity::Keypair::generate_ed25519();
+//!     let local_peer_id = PeerId::from(local_key.public());
+//!     println!("Local peer id: {:?}", local_peer_id);
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! You can already run the code above via `cargo run` which should print a
+//! different [`PeerId`](crate::PeerId) each time, given that we randomly
+//! generate the public private key pair.
+//!
+//! ## Transport
+//!
+//! Next up we need to construct a transport, after all we want to send some
+//! bytes from A to B. What exactly is a transport in the rust-libp2p world you
+//! may ask? The correct answer would be anything that implements the
+//! [`Transport`] trait. The more beginner friendly answer is any implementation
+//! of transport protocols which offer connection-oriented communication
+//! channels (e.g. TCP) as well as upgrades on top of those transports like
+//! authentication and encryption protocols.
+//!
+//! Instead of going through the many options out there combining different
+//! transport protocols and wrapping them with some authentication mechanism,
+//! let's just use the convenience function
+//! [`development_transport`](crate::development_transport) for now. It will
+//! provide us with a basic transport on top of TCP, using
+//! [`noise`](crate::noise) as an authentication mechanism.
```

Comment:
```suggestion
//! Instead of constructing a transport ourselves for this tutorial, we use the convenience
//! function [`development_transport`](crate::development_transport) that creates a TCP
//! transport with [`noise`](crate::noise) for authenticated encryption.
```